include/circt/Dialect/ESI/cosim/Endpoint.h

```diff
@@ -0,0 +1,147 @@
+  /// correctness.
+  using Blob = std::vector<uint8_t>;
+  using BlobPtr = std::shared_ptr<Blob>;
+
+  /// Construct an endpoint which knows and the type IDs in both directions.
+  Endpoint(uint64_t sendTypeId, int sendTypeMaxSize, uint64_t recvTypeId,
+           int recvTypeMaxSize);
+  ~Endpoint();
+  /// Disallow copying. There is only ONE endpoint object per logical endpoint
+  /// so copying is almost always a bug.
+  Endpoint(const Endpoint &) = delete;
+
+  uint64_t getSendTypeId() const;
+  uint64_t getRecvTypeId() const;
+
+  /// These two are used to set and unset the inUse flag, to ensure that an open
+  /// endpoint is not opened again.
+  bool setInUse();
+  void returnForUse();
+
+  /// Queue message to the simulation.
+  void pushMessageToSim(BlobPtr msg) {
+    Lock g(m);
+    toCosim.push(msg);
+  }
+
+  /// Pop from the to-simulator queue. Return true if there was a message in the
+  /// queue.
+  bool getMessageToSim(BlobPtr &msg) {
+    Lock g(m);
+    if (toCosim.size() > 0) {
+      msg = toCosim.front();
+      toCosim.pop();
+      return true;
+    }
+    return false;
+  }
+
+  /// Queue message to the RPC client.
+  void pushMessageToClient(BlobPtr msg) {
+    Lock g(m);
+    toClient.push(msg);
+  }
+
+  /// Pop from the to-RPC-client queue. Return true if there was a message in
+  /// the queue.
+  bool getMessageToClient(BlobPtr &msg) {
+    Lock g(m);
+    if (toClient.size() > 0) {
+      msg = toClient.front();
+      toClient.pop();
+      return true;
+    }
+    return false;
+  }
+
+private:
+  const uint64_t sendTypeId;
+  const uint64_t recvTypeId;
+  bool inUse;
+
+  using Lock = std::lock_guard<std::mutex>;
+
+  /// This class needs to be thread-safe. All of the mutable member variables
+  /// are protected with this object-wide lock. This may be a performance issue
+  /// in the future.
+  std::mutex m;
+  /// Message queue from RPC client to the simulation.
+  std::queue<BlobPtr> toCosim;
+  /// Message queue to RPC client from the simulation.
+  std::queue<BlobPtr> toClient;
+};
+
+/// The Endpoint registry is where Endpoints report their existence (register)
+/// and they are looked up by RPC clients.
+class EndpointRegistry {
+public:
+  /// Register an Endpoint. Creates the Endpoint object and owns it.
+  void registerEndpoint(int epId, uint64_t sendTypeId, int sendTypeMaxSize,
+                        uint64_t recvTypeId, int recvTypeMaxSize);
+
+  /// Get the specified endpoint, if it exists. Return false if it does not.
+  bool get(int epId, Endpoint *&);
+  /// Get the specified endpoint, throwing an exception if it doesn't exist.
+  /// This method is defined inline so it can be inlined at compile time.
+  /// Performance is important here since this method is used in the polling
+  /// call from the simulator.
+  Endpoint &operator[](int epId) {
+    Lock g(m);
+    auto it = endpoints.find(epId);
+    if (it == endpoints.end())
+      throw std::runtime_error("Could not locate Endpoint");
```

Comment:
Please do not use C++ exception handling.   We should build circt with -fno-exceptions.
https://llvm.org/docs/CodingStandards.html#do-not-use-rtti-or-exceptions

Reply:
Yeah, I figured I could break a few rules since it's severable and no code depends on it -- this is kind of a separate thing. Nonetheless, making this comply with the LLVM coding standards isn't a big lift.

Does LLVM have any 1:1 replacement for exceptions like the RTTI replacement?

Reply:
No, it doesn't have any drop in replacement.  It does have a variety of helpers in include/llvm/Support/ErrorHandling.h and related files.